release-22.1: ttl: replace range iterator and boundary checks with DistSQL SpanPartitions

ecwall opened the pull request:
fixes https://github.com/cockroachlabs/support/issues/1867

Backport using spans from the DistSQL planner and key decoding from 22.2
to determine table boundaries and PK values for the TTL job.

Release note (bug fix): Fix `invalid uvarint length of 9` error during TTL job.
This bug potentially affects keys with secondary tenant prefixes which includes
the serverless environment.

Release justification: Fix TTL key decoding.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, and @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFFJXq-B6DMSBsIxBPl:-NFFJXq-B6DMSBsIxBPm:b-lqdm9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b58e30ead38f505e6076e860277ae964ea9c0/pkg/sql/ttl/ttljob/ttljob.go#L713)):*
> ```Go
> 		return nil, errors.Wrapf(err, "error decoding tenant prefix of %x", key)
> 	}
> 	rKey, _, _, err = rowenc.DecodePartialTableIDIndexID(key)
> ```

I think the bug in this line is to blame for the issue, and we might consider applying a one line diff `s/key/rKey/` in the backport. That seems like a more conservative change rather than a current larger backport.


<!-- Sent from Reviewable.io -->

otan commented:
i'm fine with this or yahor's suggestion, but will not sign off as i'm not oncall for this :) !

we probably should link private repo issues in commit descriptions / PR descriptions btw.

ecwall commented on pkg/sql/ttl/ttljob/ttljob_keytodatums_test.go:
```diff
@@ -0,0 +1,33 @@
+	_, err := fmt.Sscanf(hexKey, "%x", &bytes)
+	require.NoError(t, err)
+	rKey := roachpb.RKey(bytes)
+	require.Equal(t, hexKey, fmt.Sprintf("%x", rKey))
```
I reverted the fix for now until we can get this test failing to match the prod error.
For some reason I can't figure out how to create an `RKey` that results in `hexKey` when put through `fmt.Sprintf("%x", rKey)` though so I am having trouble reproducing it:
```
Expected :2f54656e616e742f353535342f5461626c652f3132332f312f225c783163205c7865666660294c5c7866665c786265505c7862615c78646379475c7864665c78313522
Actual   :222f54656e616e742f353535342f5461626c652f3132332f312f5c225c5c783163205c5c7865666660294c5c5c7866665c5c786265505c5c7862615c5c78646379475c5c7864665c5c7831355c2222
```

ecwall replied:
The error output is not usable because of the bug that was fixed here https://github.com/cockroachdb/cockroach/pull/90727 so the test is not usable.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, and @rafiss)

___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFI1Xap0YZQcCiRYdmL):*
The release note should be more precise since I believe only multi-tenant deployments are affected which means that we should mention serverless - other deployments aren't exposed to this bug since they don't have any tenant prefix in the keys.
___
*[`pkg/sql/ttl/ttljob/ttljob_keytodatums_test.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFFGV6IlWzDyL5szsJ4-r2-21:-NFI1GPjBAdZ_oYJwI2R:b-iin2d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0f340730e93b02b0321b742a622085928a6b18b/pkg/sql/ttl/ttljob/ttljob_keytodatums_test.go#L21)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

I reverted the fix for now until we can get this test failing to match the prod error.
For some reason I can't figure out how to create an `RKey` that results in `hexKey` when put through `fmt.Sprintf("%x", rKey)` though so I am having trouble reproducing it:
```
Expected :2f54656e616e742f353535342f5461626c652f3132332f312f225c783163205c7865666660294c5c7866665c786265505c7862615c78646379475c7864665c78313522
Actual   :222f54656e616e742f353535342f5461626c652f3132332f312f5c225c5c783163205c5c7865666660294c5c5c7866665c5c786265505c5c7862615c5c78646379475c5c7864665c5c7831355c2222
```
</blockquote></details>

Note the problem is because we effectively don't strip the tenant prefix, so probably you need to somehow use a key that starts with a particular prefix (i.e. encoding of the tenant ID or something like that). 


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @yuzefovich)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFFJXq-B6DMSBsIxBPl:-NFLKtrH3sdXmprnDi8D:b-netzc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b58e30ead38f505e6076e860277ae964ea9c0/pkg/sql/ttl/ttljob/ttljob.go#L713)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think the bug in this line is to blame for the issue, and we might consider applying a one line diff `s/key/rKey/` in the backport. That seems like a more conservative change rather than a current larger backport.
</blockquote></details>

The code from 22.2 has been tested more thoroughly than the code here and the error from the support ticket has not happened in 22.2. If it has a problem then we can fix it across all versions instead of having slightly different impls that may or may not be affected.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, and @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFFJXq-B6DMSBsIxBPl:-NFU8tXj5dx5EPx4J9Cx:b-9evc99) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b58e30ead38f505e6076e860277ae964ea9c0/pkg/sql/ttl/ttljob/ttljob.go#L713)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

The code from 22.2 has been tested more thoroughly than the code here and the error from the support ticket has not happened in 22.2. If it has a problem then we can fix it across all versions instead of having slightly different impls that may or may not be affected.
</blockquote></details>

Ack, SGTM. I'll defer to Rafi for approval and proper review.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
i'm fine with the backport, but just with a bit more commenting, and also let's include the tests from https://github.com/cockroachdb/cockroach/pull/90829/ in this PR also

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-:-NFUXu5dCfSn4is40Bqy:bg081jg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, and @yuzefovich)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 409 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFUX-omEf1q_8gK9qQp:-NFUX-on5ceH_SPfvr23:bf22eda) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f77caecd2350d762491041f0526ffc60ab58959/pkg/sql/ttl/ttljob/ttljob.go#L409)):*
> ```Go
> 	}
> 
> 	distSQLPlanner := p.DistSQLPlanner()
> ```

nit: add a comment here explaining why we are using the DistSQLPlanner even though the job isn't distributed in v22.1


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @yuzefovich)

___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFI1Xap0YZQcCiRYdmL):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The release note should be more precise since I believe only multi-tenant deployments are affected which means that we should mention serverless - other deployments aren't exposed to this bug since they don't have any tenant prefix in the keys.
</blockquote></details>

Updated.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 409 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90606#-NFUX-omEf1q_8gK9qQp:-NFVUjpICrhONhMW2Iit:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f77caecd2350d762491041f0526ffc60ab58959/pkg/sql/ttl/ttljob/ttljob.go#L409)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: add a comment here explaining why we are using the DistSQLPlanner even though the job isn't distributed in v22.1
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ecwall committed:
ttl: replace range iterator and boundary checks with DistSQL SpanPartitions

fixes https://github.com/cockroachlabs/support/issues/1867

Backport using spans from the DistSQL planner and key decoding from 22.2
to determine table boundaries and PK values for the TTL job.

Release note (bug fix): Fix `invalid uvarint length of 9` error during TTL job.
This bug potentially affects keys with secondary tenant prefixes which includes
the serverless environment.
